core/src/main/scala/chisel3/properties/Property.scala

```diff
@@ -348,6 +349,47 @@ object Property {
     // This cast should be safe, because there are no members of prop.ClassType to access
     def as(prop: Property[ClassType]): Property[ClassType] with prop.ClassType =
       prop.asInstanceOf[Property[ClassType] with prop.ClassType]
+
+    def :=[T <: Class](source: Instance[T])(implicit sourceInfo: SourceInfo) = {
+      // Get the BaseModule from the source Instance.
+      val baseModule = source.getInnerDataContext.getOrElse(
+        throwException("Internal Error! Class instance did not have an associated BaseModule.")
+      )
+
+      // Get the sink Property[ClassType] expected class name.
+      val propClassName = prop.getPropertyType match {
+        case fir.ClassPropertyType(name) => name
+        case _                           => throwException("Internal Error! Property[ClassType] underlying type was not ClassPropertyType")
+      }
+
+      // Assert the sink expects the same class as the source represents.
+      if (propClassName != baseModule.name) {
```

Comment:
I don't know if there is a way to move this check to Scala's type system, but that would be great. Any thoughts @albertchen-sifive or @jackkoenig ?

Reply:
I mentioned this on the on the other comment thread but this may be the best place to discuss it. It is possible to life the String literal value to the Scala type system, but it might be very cumbersome in typical use, that being said, we can try it out. Something like:
```scala
class Class[T <: String](value: T)

val c: Class["jack"] = new Class("jack")
val d: Class["jack"] = new Class("not jack")  // this is a compile error
```

Reply:
We need the string in the `Property[ClassType]`, so I will see how it looks to try and lift that information there.

Reply:
Singleton types (which is what this is) are supported since Scala 2.13, but I will presume they are a lot more cumbersome to deal with in Scala 2 since some things users may want to do with them would required metaprogramming which is a lot more elegant in Scala 3.

Reply:
I poked at this a bit, and it seems possible, but I'm also happy to just start with a runtime check since that is straightforward.